src/nodes/chunk_dispatch/chunk_dispatch.c

```diff
@@ -290,8 +362,40 @@ chunk_dispatch_exec(CustomScanState *node)
 	/* Switch to the executor's per-tuple memory context */
 	old = MemoryContextSwitchTo(GetPerTupleMemoryContext(estate));
 
+#if PG15_GE
+	Oid partition_col_idx = InvalidOid;
+	TupleTableSlot *newslot = NULL;
+	if (dispatch->dispatch_state->mtstate->operation == CMD_MERGE)
+	{
+		partition_col_idx =
+			get_partition_attrno_from_tuple(ht->space, slot, state->colnames_list, false);
+		/* Row returned by child node does not have paritition col */
+		if (!OidIsValid(partition_col_idx))
+		{
+			List *actionStates =
+				dispatch->dispatch_state->mtstate->resultRelInfo->ri_notMatchedMergeAction;
+			ListCell *l;
+			foreach (l, actionStates)
+			{
+				MergeActionState *action = (MergeActionState *) lfirst(l);
+				CmdType commandType = action->mas_action->commandType;
+				if (commandType == CMD_INSERT)
+				{
+					/* in case received tuple has only junk attributes, then
+					 * fetch the actual values. */
+					newslot = ExecProject(action->mas_proj);
+				}
+			}
+		}
+	}
 	/* Calculate the tuple's point in the N-dimensional hyperspace */
-	point = ts_hyperspace_calculate_point(ht->space, slot);
+	if (!OidIsValid(partition_col_idx) && newslot)
+		point = ts_hyperspace_calculate_point(ht->space, newslot, NULL);
+	else
+		point = ts_hyperspace_calculate_point(ht->space, slot, state->colnames_list);
```

Comment:
What is the slot here, is it from some chunk? So this would be the case of `WHEN MATCHED THEN INSERT`. It's a little cumbersome that we now have special code for tuple conversion in the partitioning functions. I wonder if we could determine the chunk reloid here, fetch the matching chunk insert state, and then use the tuple conversion map like `hyper_to_chunk_map` to fetch the partitioning columns.

Reply:
Here slot is the tuple which ChunkDispatch node receives from its child node. This is the tuple which needs to be inserted into the target table. However if you notice from below plan output, you will see that we don't have fully projected tuple. Received slot has only 2 attributes (Output: s.temperature, t.tableoid, t.ctid). From this slot we cannot get the partitioned column value for us to decide in which chunk this tuple should be inserted into. Thus i call ExecProject() if calcuated partition slot number is invalid. Iam not doing any mapping here.

```
postgres=# explain (costs off, verbose) MERGE INTO target t
MERGE INTO target t
USING source s
ON t.time = s.time AND t.location = s.location
WHEN MATCHED AND t.location = 560076 THEN
 UPDATE SET temperature = (t.temperature + s.temperature) * 2, val = val || ' UPDATED BY MERGE'
WHEN MATCHED AND t.location = 560083 THEN
 DELETE
WHEN NOT MATCHED THEN
 INSERT VALUES ('2021-11-01 00:00:05'::timestamp with time zone, 5, 210, 'string - INSERTED BY MERGE');
                                      QUERY PLAN
------------------------------------------------------------------------------------------------
 Custom Scan (HypertableModify)
   ->  Merge on public.target t
         Merge on public.target t_1
         Merge on _timescaledb_internal._hyper_20_39_chunk t_2
         Merge on _timescaledb_internal._hyper_20_40_chunk t_3
         ->  Custom Scan (ChunkDispatch)
               ->  Merge Left Join
                     Output: s.temperature, t.tableoid, t.ctid
                     Merge Cond: ((s."time" = t."time") AND (s.location = t.location))
                     ->  Sort
                           Output: s.temperature, s."time", s.location
                           Sort Key: s."time", s.location
                           ->  Seq Scan on public.source s
                                 Output: s.temperature, s."time", s.location
                     ->  Sort
                           Output: t."time", t.location, t.tableoid, t.ctid
                           Sort Key: t."time", t.location
                           ->  Append
                                 ->  Seq Scan on public.target t_1
                                       Output: t_1."time", t_1.location, t_1.tableoid, t_1.ctid
                                 ->  Seq Scan on _timescaledb_internal._hyper_20_39_chunk t_2
                                       Output: t_2."time", t_2.location, t_2.tableoid, t_2.ctid
                                 ->  Seq Scan on _timescaledb_internal._hyper_20_40_chunk t_3
                                       Output: t_3."time", t_3.location, t_3.tableoid, t_3.ctid
(24 rows)
```